http_archive fails on windows for urls with ':'

### Description of the bug:


For 
```
http_archive(
    name = "clang-windows-x86_64",
    build_file = "//toolchains/clang:clang.BUILD",
    sha256 = "3c15cbdbc91b4921442d1660239047b3ca3dedf401119362dac9b7e803992a21",
    type = "zip",
    url = "https://chrome-infra-packages.appspot.com/dl/fuchsia/third_party/clang/windows-amd64/+/git_revision:05a34dde7008f41395312c36c95c4ebc65ff6254",
)
```
on windows 
```
bazel build "@clang-windows-x86_64//..."
Starting local Bazel server (8.5.0) and connecting to it...
Loading: 0 packages loaded
    Fetching repository @@+_repo_rules+clang-windows-x86_64; starting 39s
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'REPOSITORY_DIRECTORY:@@+_repo_rules+clang-windows-x86_64' (requested by nodes 'REPO_FILE:@@+_repo_rules+clang-windows-x86_64')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:547)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:435)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: net.starlark.java.eval.Starlark$UncheckedEvalException: InvalidPathException thrown during Starlark evaluation (repository @@+_repo_rules+clang-windows-x86_64)
        at <starlark>.download_and_extract(<builtin>:0)
        at <starlark>._http_archive_impl(C:/users/administrator/_bazel_administrator/nxaletak/external/bazel_tools/tools/build_defs/repo/http.bzl:156)
Caused by: java.nio.file.InvalidPathException: Illegal char <:> at index 133: C:\users\administrator\_bazel_administrator\nxaletak\external\+_repo_rules+clang-windows-x86_64\temp10126211350345778701\git_revision:05a34dde7008f41395312c36c95c4ebc65ff6254.zip
        at java.base/sun.nio.fs.WindowsPathParser.normalize(Unknown Source)
        at java.base/sun.nio.fs.WindowsPathParser.parse(Unknown Source)
        at java.base/sun.nio.fs.WindowsPathParser.parse(Unknown Source)
        at java.base/sun.nio.fs.WindowsPath.parse(Unknown Source)
        at java.base/sun.nio.fs.WindowsFileSystem.getPath(Unknown Source)
        at java.base/java.io.File.toPath(Unknown Source)
        at com.google.devtools.build.lib.windows.WindowsFileSystem.getAttribs(WindowsFileSystem.java:283)
        at com.google.devtools.build.lib.windows.WindowsFileSystem.stat(WindowsFileSystem.java:155)
        at com.google.devtools.build.lib.vfs.JavaIoFileSystem.getLastModifiedTime(JavaIoFileSystem.java:405)
        at com.google.devtools.build.lib.vfs.Path.getLastModifiedTime(Path.java:672)
        at com.google.devtools.build.lib.vfs.FileSystemUtils.copyFile(FileSystemUtils.java:396)
        at com.google.devtools.build.lib.bazel.repository.cache.DownloadCache.lambda$put$0(DownloadCache.java:283)
        at com.google.devtools.build.lib.bazel.repository.cache.DownloadCache.storeCacheValue(DownloadCache.java:248)
        at com.google.devtools.build.lib.bazel.repository.cache.DownloadCache.put(DownloadCache.java:282)
        at com.google.devtools.build.lib.bazel.repository.downloader.DownloadManager.downloadInExecutor(DownloadManager.java:355)
        at com.google.devtools.build.lib.bazel.repository.downloader.DownloadManager.lambda$startDownload$0(DownloadManager.java:143)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.lang.VirtualThread.run(Unknown Source)
```


### Which category does this issue belong to?

Loading API

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


_No response_

### Which operating system are you running Bazel on?


Windows

### What is the output of `bazel info release`?


release 8.5.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?


```text

```

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.


_No response_

### Have you found anything relevant by searching the web?

Can be related https://github.com/bazelbuild/bazel/issues/23729

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java
@@ -18,6 +18,7 @@ import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 
 import com.google.auth.Credentials;
+import com.google.common.base.CharMatcher;
 import com.google.common.base.MoreObjects;
 import com.google.common.base.Strings;
 import com.google.common.base.Throwables;
@@ -500,6 +501,15 @@ public class DownloadManager {
     return message.toString();
   }
 
+  // The complement of a conservative range of characters that are valid for all reasonable file
+  // systems.
+  private static final CharMatcher FS_UNSAFE_CHARS =
+      CharMatcher.inRange('a', 'z')
+          .or(CharMatcher.inRange('A', 'Z'))
+          .or(CharMatcher.inRange('0', '9'))
+          .or(CharMatcher.anyOf(".-_"))
+          .negate();
+
   private Path getDownloadDestination(URL url, Optional<String> type, Path output) {
     if (!type.isPresent()) {
       return output;
@@ -513,7 +523,9 @@ public class DownloadManager {
         basename += suffix;
       }
     }
-    return output.getRelative(basename);
+    // The basename may contain characters that aren't legal in a path with all file systems. Those
+    // characters won't matter for type determination.
+    return output.getRelative(FS_UNSAFE_CHARS.replaceFrom(basename, '_'));
   }
 
   /**

--- a/src/test/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloaderTest.java
+++ b/src/test/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpDownloaderTest.java
@@ -128,11 +128,52 @@ public class HttpDownloaderTest {
               "testCanonicalId",
               Optional.empty(),
               fs.getPath(workingDir.newFile().getAbsolutePath()),
-              eventHandler,
               Collections.emptyMap(),
               "testRepo");
 
       assertThat(new String(readFile(resultingFile), UTF_8)).isEqualTo("hello");
+    }
+  }
+
+  @Test
+  public void downloadFrom1UrlOk_specialCharInBasename() throws IOException, InterruptedException {
+    try (ServerSocket server = new ServerSocket(0, 1, InetAddress.getByName(null))) {
+      @SuppressWarnings("unused")
+      Future<?> possiblyIgnoredError =
+          executor.submit(
+              () -> {
+                try (Socket socket = server.accept()) {
+                  readHttpRequest(socket.getInputStream());
+                  sendLines(
+                      socket,
+                      "HTTP/1.1 200 OK",
+                      "Date: Fri, 31 Dec 1999 23:59:59 GMT",
+                      "Connection: close",
+                      "Content-Type: text/plain",
+                      "Content-Length: 5",
+                      "",
+                      "hello");
+                }
+                return null;
+              });
+
+      Path resultingFile =
+          download(
+              downloadManager,
+              Collections.singletonList(
+                  new URL(String.format("http://localhost:%d/arch:ve.zip", server.getLocalPort()))),
+              Collections.emptyMap(),
+              Collections.emptyMap(),
+              Optional.empty(),
+              "testCanonicalId",
+              Optional.of("zip"),
+              fs.getPath(workingDir.newFolder().getAbsolutePath()),
+              Collections.emptyMap(),
+              "testRepo");
+
+      assertThat(new String(readFile(resultingFile), UTF_8)).isEqualTo("hello");
+      assertThat(resultingFile.asFragment().getFileExtension()).isEqualTo("zip");
+      assertThat(resultingFile.asFragment().getBaseName()).doesNotContain(":");
     }
   }
 
@@ -194,7 +235,6 @@ public class HttpDownloaderTest {
               "testCanonicalId",
               Optional.empty(),
               fs.getPath(workingDir.newFile().getAbsolutePath()),
-              eventHandler,
               Collections.emptyMap(),
               "testRepo");
 
@@ -263,7 +303,6 @@ public class HttpDownloaderTest {
               "testCanonicalId",
               Optional.empty(),
               fs.getPath(workingDir.newFile().getAbsolutePath()),
-              eventHandler,
               Collections.emptyMap(),
               "testRepo");
 
@@ -334,7 +373,6 @@ public class HttpDownloaderTest {
             "testCanonicalId",
             Optional.empty(),
             outputFile,
-            eventHandler,
             Collections.emptyMap(),
             "testRepo");
         fail("Should have thrown");
@@ -679,7 +717,6 @@ public class HttpDownloaderTest {
                 "testCanonicalId",
                 Optional.empty(),
                 fs.getPath(workingDir.newFile().getAbsolutePath()),
-                eventHandler,
                 ImmutableMap.of(),
                 "testRepo"));
 
@@ -722,7 +759,6 @@ public class HttpDownloaderTest {
             "testCanonicalId",
             Optional.empty(),
             fs.getPath(workingDir.newFile().getAbsolutePath()),
-            eventHandler,
             ImmutableMap.of(),
             "testRepo");
 
@@ -771,7 +807,6 @@ public class HttpDownloaderTest {
             "testCanonicalId",
             Optional.empty(),
             fs.getPath(workingDir.newFile().getAbsolutePath()),
-            eventHandler,
             ImmutableMap.of(),
             "testRepo");
 
@@ -816,7 +851,6 @@ public class HttpDownloaderTest {
             "testCanonicalId",
             Optional.empty(),
             fs.getPath(workingDir.newFile().getAbsolutePath()),
-            eventHandler,
             ImmutableMap.of(),
             "testRepo");
 
@@ -865,7 +899,6 @@ public class HttpDownloaderTest {
             "testCanonicalId",
             Optional.empty(),
             fs.getPath(workingDir.newFile().getAbsolutePath()),
-            eventHandler,
             ImmutableMap.of(),
             "testRepo");
 
@@ -883,7 +916,6 @@ public class HttpDownloaderTest {
       String canonicalId,
       Optional<String> type,
       Path output,
-      ExtendedEventHandler eventHandler,
       Map<String, String> clientEnv,
       String context)
       throws IOException, InterruptedException {

